nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/AbstractEmailProcessor.java

```diff
@@ -0,0 +1,391 @@
+    public static final PropertyDescriptor FETCH_SIZE = new PropertyDescriptor.Builder()
+            .name("Fetch Size")
+            .description("Specify the maximum number of Messages to fetch per call to Email Server.")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .defaultValue("10")
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor SHOULD_DELETE_MESSAGES = new PropertyDescriptor.Builder()
+            .name("Delete Messages")
+            .description("Specify whether mail messages should be deleted after retrieval.")
+            .required(true)
+            .allowableValues("true", "false")
+            .defaultValue("false")
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .build();
+
+    static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("All messages that are the are successfully received from Email server and converted to FlowFiles are routed to this relationship")
+            .build();
+
+    static List<PropertyDescriptor> SHARED_DESCRIPTORS = new ArrayList<>();
+
+    static Set<Relationship> SHARED_RELATIONSHIPS = new HashSet<>();
+
+    /*
+     * Will ensure that list of PropertyDescriptors is build only once, since
+     * all other lifecycle methods are invoked multiple times.
+     */
+    static {
+        SHARED_DESCRIPTORS.add(HOST);
+        SHARED_DESCRIPTORS.add(PORT);
+        SHARED_DESCRIPTORS.add(USER);
+        SHARED_DESCRIPTORS.add(PASSWORD);
+        SHARED_DESCRIPTORS.add(FOLDER);
+        SHARED_DESCRIPTORS.add(FETCH_SIZE);
+        SHARED_DESCRIPTORS.add(SHOULD_DELETE_MESSAGES);
+
+        SHARED_RELATIONSHIPS.add(REL_SUCCESS);
+    }
+
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    protected volatile T messageReceiver;
+
+    private volatile BlockingQueue<Message> messageQueue;
+
+    private volatile String displayUrl;
+
+    private volatile ProcessSession processSession;
+
+    @OnStopped
+    public void stop(ProcessContext processContext) {
+        this.flushRemainingMessages(processContext);
+        try {
+            this.messageReceiver.destroy();
+            this.messageReceiver = null;
+        } catch (Exception e) {
+            this.logger.warn("Failure while closing processor", e);
+        }
+    }
+
+    /**
+     *
+     */
+    @Override
+    public Set<Relationship> getRelationships() {
+        return SHARED_RELATIONSHIPS;
+    }
+
+    /**
+     *
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession processSession) throws ProcessException {
+        this.initializefNecessary(context, processSession);
+
+        Message emailMessage = this.receiveMessage();
+        if (emailMessage != null) {
+            this.disposeMessage(emailMessage, context, processSession);
```

Comment:
I suspected that was the case, challenge is that disposal has a negative "to the bin" connotation that leads to confusion. (process|transfer)Message? In NiFi lingo we always transfer the message into a relationship so it may be a good candidate.